### PR TITLE
AB#11414: Tutorial and Binary Example Cleanup

### PR DESCRIPTION
**Pull request by camagallen**
In this Update:

- updated README with organization and installation info

- minor restructuring to make the library more intuitive
        * new folders called "tools", "templates", and "tutorials_and_examples"
        * new naming schema imposed

- KDD 2020 tutorial is now "archived" in the publications folder with a redirect notice to see our - now intuitively named - "Tutorial-MeasuringFairnessInBinaryClassification.ipynb".
        * allows for consistent naming conventions
        * allows for continued improvement of the tutorial if needed (as needed) without compromising the original presentation

- cleaned code for the above notebooks

- updates to binary prediction model comparison template and example (with supplemental material)
        * cleaned code
        * added explanatory content to example and especially to supplemental material

### CODE REVIEW

File: fairMLHealth/tools/model_comparison.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            return pd.DataFrame()
+        m = self.models[model_name]
+        # Cannot measure fairness without predictions
+        try:
+            y_pred = m.predict(self.X)
+        except:
+            try:
+                y_pred = m.predict(self.X.to_numpy())
+            except:
+                raise ValueError(f"Error generating predictions for {model_name}" +
```
Comment (kenviku) [inline fairMLHealth/tools/model_comparison.py:145]: I think you are looking for existence of `predict` function. As a result, you may want to change the error to whether:
1. `predict()` function exists for the model.
2. can accept inputs of type numpy array of pandas. (this is covered under the `__validate()` function)

Similar to the type checking you have under `__validate()` you can also use that to validate the first clause

File: fairMLHealth/tools/model_comparison.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        return self.__pause_validation
+
+    def __toggle_validation(self):
+        if self.__pause_validation:
+            self.__pause_validation = False
+        else:
+            self.__pause_validation = True
+
+    def measure_model(self, model_name):
+        """ Generates a report comparing fairness measures for the model_name
```
Comment (kenviku) [inline fairMLHealth/tools/model_comparison.py:126]: A good doc string example includes both parameters definition and return statements. In the return statement, while including the type of return which is in this case a pandas dataframe shall also include what is returned. For a dataframe, the expected dimensions, what is in the rows , what are the columns.

File: fairMLHealth/tools/model_comparison.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                specified
+
+            Returns:
+                a pandas dataframe
+        """
+        self.__validate()
+        if model_name not in self.models.keys():
+            print(f"Error measuring fairness: {model_name} does not appear in",
```
Comment (kenviku) [inline fairMLHealth/tools/model_comparison.py:134]: nit: For future work, convert `print` statements with python `logging`.

File: fairMLHealth/tools/reports.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    df = pd.DataFrame.from_dict(measures, orient = "index").stack().to_frame()
+    df = pd.DataFrame(df[0].values.tolist(), index = df.index)
+    df.columns = ['Value']
+    df['Value'] = df.loc[:, 'Value'].round(4)
+    df.fillna("", inplace = True)
+    return df
+
+
+
+def regression_performance(y_true, y_pred):
```
Comment (kenviku) [inline fairMLHealth/tools/reports.py:394]: 👍

File: fairMLHealth/tools/reports.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    df = pd.DataFrame.from_dict(measures, orient = "index").stack().to_frame()
+    df = pd.DataFrame(df[0].values.tolist(), index = df.index)
+    df.columns = ['Value']
+    df['Value'] = df.loc[:, 'Value'].round(4)
+    df.fillna("", inplace = True)
+    return df
+
+
+
+def classification_performance(y_true, y_pred, target_labels = None):
```
Comment (kenviku) [inline fairMLHealth/tools/reports.py:285]: 👍

File: fairMLHealth/tutorials_and_examples/Example-Template-BinaryClassificationAssessment.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
+#!/usr/bin/env python
```
Comment (kenviku) [inline fairMLHealth/tutorials_and_examples/Example-Template-BinaryClassificationAssessment.py:1]: I believe this is a notebook converted to python file? Do you intend this to be used as notebook or a python file though? It has cell numbers and format that is only supported in notebooks.

**Review by kenviku — CHANGES_REQUESTED** [review]
Thanks for the PR. In future suggest to breakdown the PRs into multiple PRs focused on specific areas. Key comments apart from the respective inline comments:
1. Change the folder package name of `fairMLHealth` to `fairmlhealth`.
2. Remove the folders on `tutorials` and `templates` from the `fairmlhealth` and move them to be parent folders of the repo.
3. Copy the contents of `tools` folder into the `fairmlhealth` folder as parent. Remove the `tools` folder.
4. Add unit tests for the `FairCompare` class. (Use `pytest`)

Finally, for any of the change request that you wish to plan/address for later PRs, open an Azure Board work-item and assign a number to the respective comment.

File: fairMLHealth/tutorials_and_examples/Example-Template-BinaryClassificationAssessment.py (hunk 6 above)
Comment (camagallen) [reply fairMLHealth/tutorials_and_examples/Example-Template-BinaryClassificationAssessment.py:1]: This line was added automatically by the converter.

File: fairMLHealth/tools/model_comparison.py (hunk 3 above)
Comment (camagallen) [reply fairMLHealth/tools/model_comparison.py:134]: Oh, good point!

File: fairMLHealth/tools/format_mimic_data.py [hunk 7]
```diff
@@ -45,16 +45,16 @@ def load_icd_ccs_xwalk(code_type):
         raise KeyError(f"Invalid code_type. Must be one of {urls.keys()}")
     s = requests.get(urls[code_type]).content
     df = pd.read_csv(io.StringIO(s.decode('utf-8')))
-    df = df.loc[:,['icd', 'ccs', 'icddesc']
-            ].rename(columns={'icd':'ICD9_CODE', 'ccs':f'{code_type}_CCS',
+    df = df.loc[:, ['icd', 'ccs', 'icddesc']
+            ].rename(columns = {'icd':'ICD9_CODE', 'ccs':f'{code_type}_CCS',
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairMLHealth/tools/format_mimic_data.py:49]: @kenviku I do have pylint  installed already, yes. It didn't flag this.

I also appear to have gotten overzealous about fixing some missing spaces. Have now switched to pylama and will fix these issues in the next PR.

File: README.md [hunk 8]
```diff
@@ -1,17 +1,43 @@
...
+
+## Installation
+Installing directly from GitHub:
+
+    python -m pip install git+https://https://github.com/KenSciResearch/fairMLHealth
+
+Installing from a local copy of the repo:
+
+    pip install <path_to_fairMLHealth_dir>
+
```
[the remark this answers is not in the document]
Comment (camagallen) [reply README.md:26]: To be completed in AB#11661

File: fairMLHealth/templates/Template-BinaryClassificationAssessment.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+# 
+# FairLearn comes with its own model comparison dashboard to allow visual comparison between models.  Note that for binary classification models the ground truth must be passed as a list.
+
+# In[ ]:
+
+
+from fairlearn.widget import FairlearnDashboard
+
+# FairLearnDashboard Note: for binary classification models, arrays must be passed as a list
+FairlearnDashboard(sensitive_features=protected_attr.to_list(),
```
[the remark this answers is not in the document]
Comment (camagallen) [reply fairMLHealth/templates/Template-BinaryClassificationAssessment.py:75]: To be completed in AB#11661

File: fairMLHealth/tools/model_comparison.py (hunk 3 above)
Comment (camagallen) [reply fairMLHealth/tools/model_comparison.py:134]: To be added in AB#11662

File: fairMLHealth/tools/model_comparison.py (hunk 1 above)
Comment (camagallen) [reply fairMLHealth/tools/model_comparison.py:145]: @kenviku the intention was to test whether the model has been fitted. Scikit has a "check_is_fitted" utility, but it appears to only work for scikit models. However we'd prefer for this object to accept any fitted model object with a predict() method... Will break this out as a separate testing function in AB#11665

File: fairmlhealth/model_comparison.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
+# -*- coding: utf-8 -*-
+"""
+Tools for measuring and comparing fairness across models
+
+"""
+# Contributors: Christine Allen <christine.allen@kensci.com>
```
Comment (kenviku) [inline fairmlhealth/model_comparison.py:6]: nit: You want to keep `camegellan`?

**Review by kenviku — APPROVED** [review]
Thanks for the changes and capturing the pending changes in work-items. Can you please add one more work-item to track for code-coverage and build pipelines?

File: fairmlhealth/model_comparison.py (hunk 10 above)
Comment (camagallen) [reply fairmlhealth/model_comparison.py:6]: @kenviku I'm sorry, I don't understand what you're asking.

**Pull request merged**